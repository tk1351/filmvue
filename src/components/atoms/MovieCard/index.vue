<script setup lang="ts">
withDefaults(
  defineProps<{
    poster_path?: string | null
    original_title?: string
    release_date?: string
  }>(),
  {
    poster_path: '',
    original_title: '',
    release_date: '',
  }
)

interface Emits {
  (eventName: 'click:link', event: Event): void
}
const emit = defineEmits<Emits>()

const handleLinkClick = (event: Event) => {
  event.preventDefault()
  emit('click:link', event)
}
</script>

<template>
  <section class="movie-card">
    <header>
      <a data-test-id="img-link" @click="handleLinkClick">
        <img class="movie-card__img" :alt="original_title" :src="poster_path" />
      </a>
    </header>
    <div class="movie-card__content">
      <h3 class="movie-card__title">
        <a data-test-id="title-link" @click="handleLinkClick">{{
          original_title
        }}</a>
      </h3>
      <p class="movie-card__date">{{ release_date }}</p>
    </div>
  </section>
</template>

<style scoped lang="scss">
.movie-card {
  width: 300px;
  height: 350px;
  border: 1px solid $main-color;
  border-radius: 30px;
  display: flex;
  flex-direction: column;
}

.movie-card__img {
  width: 100%;
  height: 220px;
  border-radius: 30px 30px 0 0;

  &:hover {
    opacity: 0.8;
  }
}

.movie-card__content {
  padding: 14px 20px;
  border-radius: 0 0 30px 30px;
  background-color: $main-color;
  flex-grow: 1;
}

.movie-card__title {
  margin-bottom: 14px;
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;

  &:hover {
    opacity: 0.8;
  }
}
</style>
