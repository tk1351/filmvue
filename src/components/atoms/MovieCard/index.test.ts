import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import MovieCard from './index.vue'

describe('Testing MovieCard Component', () => {
  it('正しくMovieCardコンポーネントが表示されていること', () => {
    const wrapper = mount(MovieCard)
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('Props', () => {
    describe('poster_path', () => {
      it('imgタグのsrc属性へ、正しく値が渡っていること', () => {
        const expectedSrc = 'expectedSrc'
        const wrapper = mount(MovieCard, {
          props: {
            poster_path: expectedSrc,
          },
        })
        const image = wrapper.find('img')
        expect(image.attributes('src')).toBe(expectedSrc)
      })
    })

    describe('original_title', () => {
      it('imgのalt属性へ、正しく値が渡っていること', () => {
        const expectedTitle = 'expectedTitle'
        const wrapper = mount(MovieCard, {
          props: {
            original_title: expectedTitle,
          },
        })
        const image = wrapper.find('img')
        expect(image.attributes('alt')).toBe(expectedTitle)
      })
      it('h3タグの中に、正しく値が渡っていること', () => {
        const expectedTitle = 'expectedTitle'
        const wrapper = mount(MovieCard, {
          props: {
            original_title: expectedTitle,
          },
        })
        expect(wrapper.find('h3').text()).toContain(expectedTitle)
      })
    })

    describe('release_date', () => {
      it('pタグに、正しく値が渡っていること', () => {
        const expectedDate = 'expectedDate'
        const wrapper = mount(MovieCard, {
          props: {
            release_date: expectedDate,
          },
        })
        expect(wrapper.find('p').text()).toContain(expectedDate)
      })
    })
  })

  describe('Event', () => {
    it('画像のリンクをクリックした時、click:linkイベントが1度実行されること', async () => {
      const wrapper = mount(MovieCard)
      await wrapper.find("[data-test-id='img-link']").trigger('click')
      expect(wrapper.emitted('click:link')).toHaveLength(1)
    })
    it('タイトルのリンクをクリックした時、click:linkイベントが1度実行されること', async () => {
      const wrapper = mount(MovieCard)
      await wrapper.find("[data-test-id='title-link']").trigger('click')
      expect(wrapper.emitted('click:link')).toHaveLength(1)
    })
  })
})
