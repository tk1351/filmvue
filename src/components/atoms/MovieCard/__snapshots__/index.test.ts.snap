// Vitest Snapshot v1

exports[`Testing MovieCard Component > 正しくMovieCardコンポーネントが表示されていること 1`] = `
<section
  class="movie-card"
  data-v-61cece96=""
>
  <header
    data-v-61cece96=""
  >
    <a
      data-test-id="img-link"
      data-v-61cece96=""
    >
      <img
        alt=""
        class="movie-card__img"
        data-v-61cece96=""
        src=""
      />
    </a>
  </header>
  <div
    class="movie-card__content"
    data-v-61cece96=""
  >
    <h3
      class="movie-card__title"
      data-v-61cece96=""
    >
      <a
        data-test-id="title-link"
        data-v-61cece96=""
      />
    </h3>
    <p
      class="movie-card__date"
      data-v-61cece96=""
    />
  </div>
</section>
`;
